### PR TITLE
Marc/abstract mediation

### PR DESCRIPTION
**Pull request by fogAndWhisky**
Creates an AbstractMediationBinder, abstracting non-MonoBehaviour aspects out of the MediationBinder. This allows a measure of Unit Testing in previsously untestable code, and makes it possible for MediationBinder and the upcoming EditorMediationBinder to be DRYer and more flexible.

### CODE REVIEW

File: StrangeIoC/scripts/strange/extensions/mediation/impl/MediationBinder.cs [hunk 1]
```diff
@@ -17,212 +17,87 @@
...
-					MonoBehaviour mono = view as MonoBehaviour;
-					IMediator mediator = mono.GetComponent(mediatorType) as IMediator;
-					if (mediator != null)
-					{
-						removeMediator(mediator);
-					}
-				}
+				DestroyMediator (mediator);
 			}
+			return DestroyMediator (mediator);
```
Comment (WillzZz) [inline StrangeIoC/scripts/strange/extensions/mediation/impl/MediationBinder.cs]: DestroyMediator 2x if we have a mediator?

File: StrangeIoC/scripts/strange/extensions/mediation/impl/MediationBinder.cs [hunk 2]
```diff
@@ -17,212 +17,87 @@
...
-		}
-
-		/// Creates and registers one or more Mediators for a specific View instance.
-		/// Takes a specific View instance and a binding and, if a binding is found for that type, creates and registers a Mediator.
-		virtual protected void mapView(IView view, IMediationBinding binding)
-		{
-			Type viewType = view.GetType();
-
-			if (bindings.ContainsKey(viewType))
+			bool isTrueMediator = mediatorType.IsAssignableFrom (typeof(IMediator));
```
Comment (WillzZz) [inline StrangeIoC/scripts/strange/extensions/mediation/impl/MediationBinder.cs]: This looks backwards. Fixing on my local branch. Just thought I'd point it out for future use. interfaceType.isassignablefrom(myType)

File: StrangeIoC/scripts/strange/extensions/mediation/impl/AbstractMediationBinder.cs [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+				{
+					throw new BinderException ("A runtime Mediation Binding has resolved to null. Did you forget to register its fully-qualified name?\n View:" + key, BinderExceptionType.RUNTIME_NULL_VALUE);
+				}
+				if (binding == null)
+				{
+					binding = Bind (keyType);
+				}
+				else
+				{
+					binding = binding.Bind (keyType);
```
Comment (WillzZz) [inline StrangeIoC/scripts/strange/extensions/mediation/impl/AbstractMediationBinder.cs]: What is the case for this? How do we have multiple keyList here? This line isn't covered so I was going to test it.

File: StrangeIoC/scripts/strange/framework/impl/Binder.cs [hunk 4]
```diff
@@ -349,43 +349,51 @@ virtual public void ConsumeBindings(string jsonString)
...
 			{
-				Dictionary<string, object> item = list[a] as Dictionary<string, object>;
-				List<object> keyList;
-				List<object> valueList;
-				IBinding binding = null;
+				ConsumeItem(list[a] as Dictionary<string, object>, testBinding);
+			}
+		}
+
+		virtual protected IBinding ConsumeItem(Dictionary<string, object> item, IBinding testBinding)
```
Comment (WillzZz) [inline StrangeIoC/scripts/strange/framework/impl/Binder.cs]: Very well covered but I think some of these methods are worthy of comment/summary.

**Comment by WillzZz** [comment]
Tests look good, I think it's ready to move to tip pretty soon.

**Comment by fogAndWhisky** [comment]
Looking at this today.

File: StrangeIoC/scripts/strange/extensions/mediation/impl/AbstractMediationBinder.cs (hunk 3 above)
Comment (fogAndWhisky) [inline StrangeIoC/scripts/strange/extensions/mediation/impl/AbstractMediationBinder.cs]: Nah, it's a mistake. The default keyConstraint is for a Binding is BindingConstraintType.ONE, and MediationBinding doesn't override that. So this condition can never arise.

Comment (fogAndWhisky) [inline StrangeIoC/scripts/strange/extensions/mediation/impl/AbstractMediationBinder.cs]: It does point out an interesting Exception that we're NOT throwing, though:

    Bind<This>().Bind<AndThis>().To<That>();

is an error in a BindingConstraint.ONE situation. But we don't throw. In most cases, we probably handle this at the Binder level...just wondering aloud if the Binding itself should make some noise if a dev does this. If not, should the AbstractMediationBinder throw an Exception on this line? THAT we could test. :)

**Comment by fogAndWhisky** [comment]
Pushed the documentation changes you asked for. Still considering the best answer for the 'Double Bind' case. I feel like we should throw an error in Binding.cs if someone attempts to 'Bind' or 'To' or 'Name' in a way that violates the BindingConstraint rules. But if so, we can only do it for strong bindings, not weak ones, right?

**Comment by fogAndWhisky** [comment]
So I've dug more deeply into the BindingConstraint rules this morning. I guess I had forgotten that we deliberately allow SemiBindings to be overwritten. As such, we don't really throw the sort of error I was contemplating...rather, we simply overwrite. Arguably this is something we should consider doing...but as it's not especially relevant to this PR, I've simply corrected my mistake and we'll move on, pending your approval of this PR.

**Comment by WillzZz** [comment]
Alright I merged but I want to make sure the proj file isn't mucked up after my merge. Test it out then we'll click the big green button.

**Comment by WillzZz** [comment]
Tested basics, again. Merging.

**Pull request merged**